src/Microsoft.AspNet.Security/Infrastructure/AuthenticationHandler.cs

```diff
@@ -307,7 +310,18 @@ protected virtual void ApplyResponseCore()
         protected virtual async Task ApplyResponseCoreAsync()
         {
             await ApplyResponseGrantAsync();
-            await ApplyResponseChallengeAsync();
+
+            // If authenticate was called and the the status is still 401, authZ failed so set 403 and stop
+            // REVIEW: Does this need to ensure that there's the 401 is challenge for this auth type?
+            if (Response.StatusCode == 401 && _authenticateCalled)
```

Comment:
One related note: It's quite strange for cookie auth to issue 403s if it's set up to handle 401s by redirecting to a login page. Wouldn't it make more sense for it to redirect to an Access Denied page instead?